global_buffer/global_buffer_bs_generator.py

```diff
@@ -0,0 +1,214 @@
+                             f"configuration registers to the feature")
+
+        reg_name = f"{self.name()}_{name}"
+        addr = (self.core.core_id << (self.feature_addr_width +
+                                      self.reg_addr_width +
+                                      self.byte_offset)) | \
+               (self.feature_id << (self.reg_addr_width +
+                                    self.byte_offset)) | \
+               (self.num_registers << self.byte_offset)
+        self.registers[f"{name}"] = {"name": reg_name, "addr": addr,
+                                     "range": width, "default": default_value,
+                                     "value": default_value}
+        self.num_registers = self.num_registers + 1
+
+    @abstractmethod
+    def name(self):
+        pass
+
+
+class IOController(GlbCore):
+    def __init__(self,
+                 glb: "Glb",
+                 core_id: int,
+                 num_io: int):
+        super().__init__(glb, core_id)
+        self.num_io = num_io
+
+        for i in range(self.num_io):
+            _io_address_generator = IOAddrGen(self, i)
+            self.add_feature(_io_address_generator)
+
+    def name(self):
+        return f"GLB_IOController"
+
+
+class IOAddrGen(GlbFeature):
+    def __init__(self,
+                 core: "GlbCore",
+                 feature_id: int):
+        super().__init__(core, feature_id)
+
+        # reg_addr is determined by the order you add config here
+        switch_size = int(math.ceil(self.glb.num_banks / self.core.num_io))
+        self.add_config("mode", 2)
+        self.add_config("start_addr", 32)
+        self.add_config("num_words", 32)
+        self.add_config("switch_sel", 4)
+        self.add_config("done_delay", switch_size)
+        self.add_config("done_gate", 1)
+
+    def name(self):
+        return f"{self.core.name()}_IOAddrGen_{self.feature_id}"
+
+
+class ParCfgController(GlbCore):
+    def __init__(self,
+                 glb: "Glb",
+                 core_id: int,
+                 num_cfg: int):
+        super().__init__(glb, core_id)
+        self.num_cfg = num_cfg
+
+        for i in range(self.num_cfg):
+            _par_cfg_address_generator = ParCfgAddrGen(self, i)
+            self.add_feature(_par_cfg_address_generator)
+
+    def name(self):
+        return f"GLB_ParCfgController"
+
+
+class ParCfgAddrGen(GlbFeature):
+    def __init__(self,
+                 core: "GlbCore",
+                 feature_id: int):
+        super().__init__(core, feature_id)
+
+        # reg_addr is determined by the order you add config here
+        switch_size = int(math.ceil(self.glb.num_banks / self.core.num_cfg))
+        self.add_config("start_addr", 32)
+        self.add_config("num_words", 32)
+        self.add_config("switch_sel", switch_size)
+
+    def name(self):
+        return f"{self.core.name()}_ParCfgAddrGen_{self.feature_id}"
+
+
+def GlbBSGenerator(glb: "Glb", collateral):
```

Comment:
Consider refactor this function to the `Glb` object. This is a functionality of the global buffer. You can refer to how `Interconnect` class handles that.

Reply:
okay.
Will refactor this function to Glb object.